src/mlpack/core/tree/rectangle_tree/r_plus_tree_split_impl.hpp

```diff
@@ -0,0 +1,342 @@
+  // Update the number of descandants.
+  treeOne->numDescendants = treeOne->Count();
+  treeTwo->numDescendants = treeTwo->Count();
+
+  assert(treeOne->Count() <= treeOne->MaxLeafSize());
+  assert(treeTwo->Count() <= treeTwo->MaxLeafSize());
+  
+  assert(tree->Count() == treeOne->Count() + treeTwo->Count());
+  assert(treeOne->Bound()[cutAxis].Hi() < treeTwo->Bound()[cutAxis].Lo());
+}
+
+template<typename SplitPolicyType,
+         template<typename> class SweepType>
+template<typename TreeType>
+void RPlusTreeSplit<SplitPolicyType, SweepType>::SplitNonLeafNodeAlongPartition(
+    TreeType* tree,
+    TreeType* treeOne,
+    TreeType* treeTwo,
+    const size_t cutAxis,
+    const typename TreeType::ElemType cut)
+{
+  // Split the auxiliary information.
+  tree->AuxiliaryInfo().SplitAuxiliaryInfo(treeOne, treeTwo, cutAxis, cut);
+
+  // Insert children into the corresponding subtree.
+  for (size_t i = 0; i < tree->NumChildren(); i++)
+  {
+    TreeType* child = tree->children[i];
+    int policy = SplitPolicyType::GetSplitPolicy(*child, cutAxis, cut);
+
+    if (policy == SplitPolicyType::AssignToFirstTree)
+    {
+      InsertNodeIntoTree(treeOne, child);
+      child->Parent() = treeOne;
+    }
+    else if (policy == SplitPolicyType::AssignToSecondTree)
+    {
+      InsertNodeIntoTree(treeTwo, child);
+      child->Parent() = treeTwo;
+    }
+    else
+    {
+      // The child should be split (i.e. the partition divides its bound).
+      TreeType* childOne = new TreeType(treeOne);
+      TreeType* childTwo = new TreeType(treeTwo);
+      treeOne->MinLeafSize() = 0;
+      treeOne->MinNumChildren() = 0;
+      treeTwo->MinLeafSize() = 0;
+      treeTwo->MinNumChildren() = 0;
+
+      // Propagate the split downward.
+      if (child->IsLeaf())
+        SplitLeafNodeAlongPartition(child, childOne, childTwo, cutAxis, cut);
+      else
+        SplitNonLeafNodeAlongPartition(child, childOne, childTwo, cutAxis, cut);
+
+      InsertNodeIntoTree(treeOne, childOne);
+      InsertNodeIntoTree(treeTwo, childTwo);
+
+      child->SoftDelete();
+    }
+  }
+
+  assert(treeOne->NumChildren() + treeTwo->NumChildren() != 0);
+
+  // Add a fake subtree if one of the subtrees is empty.
+  if (treeOne->NumChildren() == 0)
+    AddFakeNodes(treeTwo, treeOne);
+  else if (treeTwo->NumChildren() == 0)
+    AddFakeNodes(treeOne, treeTwo);
+
+  assert(treeOne->NumChildren() <= treeOne->MaxNumChildren());
+  assert(treeTwo->NumChildren() <= treeTwo->MaxNumChildren());
+}
+
+template<typename SplitPolicyType,
+         template<typename> class SweepType>
+template<typename TreeType>
+void RPlusTreeSplit<SplitPolicyType, SweepType>::
+AddFakeNodes(const TreeType* tree, TreeType* emptyTree)
+{
+  size_t numDescendantNodes = tree->TreeDepth() - 1;
+
+  TreeType* node = emptyTree;
+  for (size_t i = 0; i < numDescendantNodes; i++)
+  {
+    TreeType* child = new TreeType(node);
+    node->children[node->NumChildren()++] = child;
+
+    node = child;
+  }
+}
```

Comment:
Just to be sure I am understanding correctly, this function is used to make sure that the tree has equivalent maximum depth in every branch?